### DISCUSSION TITLE
Keep apps opened on browser refresh

### DISCUSSION THREAD

**Discussion by mtchuenten** [os-js/OS.js#830] [General]
Hello all,

In my OS.js installation, when I log in and launch other apps, and then refresh the browser, I'm presented with the log in form again. 

Is there a way to prevent that, such that my apps remain opened even when I refresh the browser?

Thanks.

**Comment by andersevenrud** [comment]
You have to save session and log out instead of doing a refresh.

**Reply by mtchuenten** [reply]
Thank you for your reply @andersevenrud. 

I already use the _Save Session & Logout_ option, and that works fine. I was merely anticipating in case a user refreshes their browser instead of using that option. You know, users can be very unpredictable :sweat_smile:. Anyway, that's not a big deal really.

**Reply by andersevenrud** [reply]
It's possible to hook into the event bus to do automatic session saves whenever a window/app is manipulated somehow.

So maybe this could be transformed into an issue instead, because it should be fairly trivial to make this a configuration option 🤔

**Reply by andersevenrud** [reply]
Just wrote this quick and dirty example on how it might be possible right now by adding your own service:

**Not tested** 😅 

```js
import { debounce } from 'lodash'

export class AutoSaveSessionServiceProvider {
	constructor(core) {
		this.core = core

		this.onChange = debounce(() => core.make('osjs/session').save(), 500)

		this.listeners = [
			'osjs/window:destroy',
			'osjs/window:create',
			'osjs/window:change',
			'osjs/application:create',
			'osjs/application:destroy'
		]
	}

	destroy() {
		this.listeners.forEach(n => this.core.off(n, this.onChange))
		this.onChange = () => { /* Noop */ }
	}

	async init() {
		this.listeners.forEach(n => this.core.on(n, this.onChange))
	}
}

```

**Reply by mtchuenten** [reply]
Thank you for your response @andersevenrud.

Please let me know how to use it effectively, I'm probably missing something somewhere, reason why when I refresh the browser I'm still presented with the login form.

This is what I did:

1. Saved your code above in src/autoSaveSessionServiceProvider.js
2. Updated src/client/index.js as shown below:
```
import {
  Core,
  CoreServiceProvider,
  DesktopServiceProvider,
  VFSServiceProvider,
  NotificationServiceProvider,
  SettingsServiceProvider,
  AuthServiceProvider,
} from "@osjs/client";
import { AutoSaveSessionServiceProvider } from "../autoSaveSessionServiceProvider";

import { PanelServiceProvider } from "@osjs/panels";
import { GUIServiceProvider } from "@osjs/gui";
import { DialogServiceProvider } from "@osjs/dialogs";
import config from "./config.js";
import "./index.scss";
import myAdapter from "./auth/my-auth.js";

const init = () => {
  const osjs = new Core(config, {});

  // Register your service providers
  osjs.register(CoreServiceProvider);
  osjs.register(AutoSaveSessionServiceProvider);
  osjs.register(DesktopServiceProvider);
  osjs.register(VFSServiceProvider);
  osjs.register(NotificationServiceProvider);
  osjs.register(SettingsServiceProvider, { before: true });
  osjs.register(AuthServiceProvider, {
    before: true,
    args: { adapter: myAdapter }
  });
  osjs.register(PanelServiceProvider);
  osjs.register(DialogServiceProvider);
  osjs.register(GUIServiceProvider);
  osjs.boot();
};

window.addEventListener("DOMContentLoaded", () => init());
```
3. Updated src/server/index.js as shown below:
```
const {
  Core,
  CoreServiceProvider,
  PackageServiceProvider,
  VFSServiceProvider,
  AuthServiceProvider,
  SettingsServiceProvider,
} = require("@osjs/server");
const {
  AutoSaveSessionServiceProvider,
} = require("../autoSaveSessionServiceProvider");

const config = require("./config.js");
const osjs = new Core(config, {});

osjs.register(CoreServiceProvider, { before: true });
osjs.register(AutoSaveSessionServiceProvider, { before: true });
osjs.register(PackageServiceProvider);
osjs.register(VFSServiceProvider);
osjs.register(AuthServiceProvider);
osjs.register(SettingsServiceProvider);

const shutdown = (signal) => (error) => {
  if (error instanceof Error) {
    console.error(error);
  }

  osjs.destroy(() => process.exit(signal));
};

process.on("SIGTERM", shutdown(0));
process.on("SIGINT", shutdown(0));
process.on("exit", shutdown(0));

osjs.boot().catch(shutdown(1));
```
4. Ran _npm run build_

Regards.

**Reply by andersevenrud** [reply]
It's a client-only service.

You only need `osjs.register(AutoSaveSessionServiceProvider);` in your client bootstrap.

> reason why when I refresh the browser I'm still presented with the login form.

Yes. This only automatically saves the session data. It does not keep your session alive.

See: https://github.com/os-js/osjs-client/issues/88 and https://github.com/os-js/OS.js/issues/761#issuecomment-808608051

**Reply by mtchuenten** [reply]
> It's a client-only service.
> 
> You only need osjs.register(AutoSaveSessionServiceProvider); in your client bootstrap.

I see. I tried client only, didn't work, then I also involved the server.

> Yes. This only automatically saves the session data. It does not keep your session alive.
> 
> See: https://github.com/os-js/osjs-client/issues/88 and https://github.com/os-js/OS.js/issues/761#issuecomment-808608051

Ok, thanks. Will check.
